gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.dist;
+
+import static com.palantir.logsafe.Preconditions.checkArgument;
+
+import com.palantir.logsafe.SafeArg;
+import com.palantir.sls.versions.OrderableSlsVersion;
+import com.palantir.sls.versions.VersionComparator;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+public final class RecommendedProductDependencyMerger {
+    private RecommendedProductDependencyMerger() { }
+
+    public static RecommendedProductDependency mergeRecommendedProductDependencies(
+            RecommendedProductDependency dep1, RecommendedProductDependency dep2) {
+        // Ensure they are valid
+        Arrays.asList(dep1, dep2).forEach(RecommendedProductDependency::isValid);
+        checkArgument(
+                dep1.getProductGroup().equals(dep2.getProductGroup()),
+                "Product groups differ",
+                SafeArg.of("dep1ProductGroup", dep1.getProductGroup()),
```

Comment:
I'm not sure how useful the safe-logging preconditions here are, as we're just going to use gradle's SLF4J implementation... I think this means we need `{}` tokens in the message (or maybe `%s` I can never remember)

Reply:
Let's switch to the guava ones for non-server usage

https://github.com/palantir/safe-logging/blob/develop/preconditions/src/main/java/com/palantir/logsafe/exceptions/SafeIllegalStateException.java